<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Source: setup.js</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Source: setup.js</h1>

    



    
    <section>
        <article>
            <pre class="prettyprint source linenums"><code>/* eslint-disable no-global-assign */

/**
 * @file   Replace global async methods with Promise-based ones and it adds async features to NodeJS's async.
 * @author Manuel Spagnolo.
 * @since  0.0.1
 */

const { isNumber } = require('./_utils');
const assert = require('assert');

/**
 * A reference to NodeJS's core &lt;tt>setTimeout&lt;/tt> and &lt;tt>setInterval&lt;/tt>
 * @constant
 * @type {object}
 * @private
 */
const oldies = {setTimeout, setInterval};

/**
 * Overrides NodeJS's core &lt;tt>setTimeout&lt;/tt> replacing it with a Promise-based one.
 *
 * It's meant to work exactly like the old &lt;tt>setTimeout&lt;/tt> and they should be indistinguishable.
 * @since 0.0.1
 *
 * @async
 *
 * @param {function} callback   The callback to be executed after
 *                              &lt;tt>timeout&lt;/tt> ms
 * @param {number} timeout      How many ms to wait before calling
 *                              &lt;tt>timeout&lt;/tt>
 *
 * @returns {Promise}
 */
setTimeout = (callback, timeout) => new Promise((resolve, reject) => {
  if (!isNumber(timeout) &amp;&amp; timeout >= 0) {
    reject(new TypeError('Timeout should be a positive number!'));
  }

  oldies.setTimeout(() => {
    try {
      callback();
      resolve();
    } catch (e) {
      reject(e);
    }
  }, timeout);
});

/**
 * Overrides NodeJS's core &lt;tt>setInterval&lt;/tt> replacing it with a Promise-based one.
 *
 * It's meant to work exactly like the old &lt;tt>setInterval&lt;/tt> and they should
 * be indistinguishable.
 * @since 0.0.1
 *
 * @async
 *
 * @param {function} callback   The callback to be executed every &lt;tt>interval&lt;/tt> ms
 * @param {number} interval     How many ms to wait before calling &lt;tt>interval&lt;/tt>
 *
 * @returns {Promise}
 */
setInterval = (callback, interval) => new Promise((resolve, reject) => {
  if (!isNumber(interval) &amp;&amp; interval >= 0) {
    reject(new TypeError('Interval should be a positive number!'));
  }

  oldies.setInterval(() => {
    try {
      callback();
      resolve();
    } catch (e) {
      reject(e);
    }
  }, interval);
});

/** @namespace assert */
/**
 * Adds the ability to check if async functions throw to NodeJS's &lt;tt>assert&lt;/tt> library.
 *
 * Has the same signature as
 * [assert.throws]{@link http://nodejs.org/api/assert.html#assert_assert_throws_block_error_message}
 *
 * @function assert~throwsAsync
 *
 * @since 0.0.1
 *
 * @async
 *
 * @param {function}                  block     Function to check
 * @param {constructor|string|RegExp} [error]   Expected thrown error
 * @param {string}                    [message] Message given by &lt;tt>AssertionError&lt;/tt>
 *
 * @returns {Promise}
 */
assert.throwsAsync = async (block, error, message) => {
  let f = () => { };
  try {
    await block();
  } catch (e) {
    f = () => {
      throw e;
    };
  } finally {
    assert.throws(f, error, message);
  }
};

/**
 * Adds the ability to check if async functions does not throw to NodeJS's &lt;tt>assert&lt;/tt> library.
 *
 * Has the same signature as
 * [assert.doesNotThrow]{@link https://nodejs.org/api/assert.html#assert_assert_doesnotthrow_block_error_message}
 *
 * @function assert~doesNotThrowAsync
 *
 * @since 0.0.1
 *
 * @async
 *
 * @param {function}                  block     Function to check
 * @param {constructor|string|RegExp} [error]   Expected thrown error
 * @param {string}                    [message] Message given by &lt;tt>AssertionError&lt;/tt>
 *
 * @returns {Promise}
 */
assert.doesNotThrowAsync = async (block, error, message) => {
  let f = () => { };
  try {
    await block();
  } catch (e) {
    f = () => {
      throw e;
    };
  } finally {
    assert.doesNotThrow(f, error, message);
  }
};
/* eslint-enable no-global-assign */
</code></pre>
        </article>
    </section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="assert.html">assert</a></li></ul><h3>Global</h3><ul><li><a href="global.html#setInterval">setInterval</a></li><li><a href="global.html#setTimeout">setTimeout</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Tue Jan 09 2018 19:29:24 GMT+0100 (W. Europe Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>
